============================================================
HNS-OAuth-DFS
============================================================

[ERROR] org.apache.hadoop.fs.azurebfs.ITestAzureBlobFileSystemRename.testRenamePathRetryIdempotency  Time elapsed: 1.324 s  <<< ERROR!

[WARNING] Tests run: 174, Failures: 0, Errors: 0, Skipped: 3
[ERROR] Tests run: 798, Failures: 0, Errors: 1, Skipped: 154
[WARNING] Tests run: 157, Failures: 0, Errors: 0, Skipped: 7
[WARNING] Tests run: 272, Failures: 0, Errors: 0, Skipped: 23

============================================================
HNS-SharedKey-DFS
============================================================

[ERROR] org.apache.hadoop.fs.azurebfs.ITestAzureBlobFileSystemRename.testRenamePathRetryIdempotency  Time elapsed: 0.978 s  <<< ERROR!

[WARNING] Tests run: 174, Failures: 0, Errors: 0, Skipped: 4
[ERROR] Tests run: 801, Failures: 0, Errors: 1, Skipped: 107
[WARNING] Tests run: 157, Failures: 0, Errors: 0, Skipped: 7
[WARNING] Tests run: 272, Failures: 0, Errors: 0, Skipped: 10

============================================================
NonHNS-SharedKey-DFS
============================================================

[WARNING] Tests run: 174, Failures: 0, Errors: 0, Skipped: 10
[WARNING] Tests run: 640, Failures: 0, Errors: 0, Skipped: 212
[WARNING] Tests run: 157, Failures: 0, Errors: 0, Skipped: 8
[WARNING] Tests run: 272, Failures: 0, Errors: 0, Skipped: 11

============================================================
AppendBlob-HNS-OAuth-DFS
============================================================

[ERROR] org.apache.hadoop.fs.azurebfs.ITestAzureBlobFileSystemRename.testRenamePathRetryIdempotency  Time elapsed: 1.17 s  <<< ERROR!

[WARNING] Tests run: 174, Failures: 0, Errors: 0, Skipped: 3
[ERROR] Tests run: 798, Failures: 0, Errors: 1, Skipped: 165
[WARNING] Tests run: 134, Failures: 0, Errors: 0, Skipped: 8
[WARNING] Tests run: 272, Failures: 0, Errors: 0, Skipped: 23

============================================================
NonHNS-SharedKey-Blob
============================================================

[WARNING] Tests run: 174, Failures: 0, Errors: 0, Skipped: 10
[WARNING] Tests run: 643, Failures: 0, Errors: 0, Skipped: 141
[WARNING] Tests run: 157, Failures: 0, Errors: 0, Skipped: 3
[WARNING] Tests run: 272, Failures: 0, Errors: 0, Skipped: 11

============================================================
NonHNS-OAuth-DFS
============================================================

[WARNING] Tests run: 174, Failures: 0, Errors: 0, Skipped: 10
[WARNING] Tests run: 637, Failures: 0, Errors: 0, Skipped: 214
[WARNING] Tests run: 157, Failures: 0, Errors: 0, Skipped: 8
[WARNING] Tests run: 272, Failures: 0, Errors: 0, Skipped: 24

============================================================
NonHNS-OAuth-Blob
============================================================

[WARNING] Tests run: 174, Failures: 0, Errors: 0, Skipped: 10
[WARNING] Tests run: 640, Failures: 0, Errors: 0, Skipped: 143
[WARNING] Tests run: 157, Failures: 0, Errors: 0, Skipped: 3
[WARNING] Tests run: 272, Failures: 0, Errors: 0, Skipped: 24

============================================================
AppendBlob-NonHNS-OAuth-Blob
============================================================

[WARNING] Tests run: 174, Failures: 0, Errors: 0, Skipped: 10
[WARNING] Tests run: 638, Failures: 0, Errors: 0, Skipped: 161
[WARNING] Tests run: 134, Failures: 0, Errors: 0, Skipped: 4
[WARNING] Tests run: 272, Failures: 0, Errors: 0, Skipped: 24

============================================================
HNS-Oauth-DFS-IngressBlob
============================================================

[ERROR] org.apache.hadoop.fs.azurebfs.ITestAzureBlobFileSystemRename.testRenamePathRetryIdempotency  Time elapsed: 1.464 s  <<< ERROR!

[WARNING] Tests run: 174, Failures: 0, Errors: 0, Skipped: 3
[ERROR] Tests run: 672, Failures: 0, Errors: 1, Skipped: 161
[WARNING] Tests run: 157, Failures: 0, Errors: 0, Skipped: 7
[WARNING] Tests run: 272, Failures: 0, Errors: 0, Skipped: 23

============================================================
NonHNS-OAuth-DFS-IngressBlob
============================================================

[WARNING] Tests run: 174, Failures: 0, Errors: 0, Skipped: 10
[WARNING] Tests run: 637, Failures: 0, Errors: 0, Skipped: 212
[WARNING] Tests run: 157, Failures: 0, Errors: 0, Skipped: 8
[WARNING] Tests run: 272, Failures: 0, Errors: 0, Skipped: 24

--- a/hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java
+++ b/hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java
@@ -1300,6 +1300,8 @@ public class AzureBlobFileSystemStore implements Closeable, ListingSupport {
 
     if (listingClient instanceof AbfsBlobClient) {
       fileStatuses.addAll(ListUtils.getUniqueListResult(fileStatusList));
+      LOG.debug("ListBlob API returned a total of {} elements including duplicates."
+          + "Number of unique Elements are {}", fileStatusList.size(), fileStatuses.size());
     } else {
       fileStatuses.addAll(fileStatusList);
     }

--- a/hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java
+++ b/hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java
@@ -40,13 +40,11 @@ import java.util.Date;
 import java.util.HashMap;
 import java.util.HashSet;
 import java.util.Hashtable;
-import java.util.Iterator;
 import java.util.List;
 import java.util.Locale;
 import java.util.Map;
 import java.util.Optional;
 import java.util.Set;
-import java.util.TreeMap;
 import java.util.WeakHashMap;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.ExecutorService;
@@ -120,6 +118,7 @@ import org.apache.hadoop.fs.azurebfs.utils.Base64;
 import org.apache.hadoop.fs.azurebfs.utils.CRC64;
 import org.apache.hadoop.fs.azurebfs.utils.DateTimeUtils;
 import org.apache.hadoop.fs.azurebfs.utils.EncryptionType;
+import org.apache.hadoop.fs.azurebfs.utils.ListUtils;
 import org.apache.hadoop.fs.azurebfs.utils.TracingContext;
 import org.apache.hadoop.fs.azurebfs.utils.UriUtils;
 import org.apache.hadoop.fs.impl.BackReference;
@@ -1300,53 +1299,12 @@ public class AzureBlobFileSystemStore implements Closeable, ListingSupport {
     } while (shouldContinue);
 
     if (listingClient instanceof AbfsBlobClient) {
-      fileStatuses.addAll(removeDuplicates(fileStatusList));
+      fileStatuses.addAll(ListUtils.getUniqueListResult(fileStatusList));
     } else {
       fileStatuses.addAll(fileStatusList);
     }
 
     return continuation;
-  }
-
-  /**
-   * This is to handle duplicate listing entries returned by Blob Endpoint for
-   * implicit paths that also has a marker file created for them.
-   * This will retain the entry corresponding to the marker file
-   * and remove the BlobPrefix entry corresponding to implicit directory.
-   * @param fileStatusList the list of file statuses to be rectified.
-   */
-  private List<FileStatus> removeDuplicates(List<FileStatus> fileStatusList) {
-    if (fileStatusList.isEmpty()) {
-      return fileStatusList;
-    }
-    List<FileStatus> rectifiedFileStatusList = new ArrayList<>();
-    Iterator<FileStatus> iterator = fileStatusList.iterator();
-    FileStatus curr = iterator.next();
-    TreeMap<String, FileStatus> nameToEntryMap = new TreeMap<>();
-    String prefix = curr.getPath().getName();
-    nameToEntryMap.put(prefix, curr);
-    rectifiedFileStatusList.add(curr);
-    while (iterator.hasNext()) {
-      FileStatus next = iterator.next();
-      if (next.getPath().getName().startsWith(prefix)) {
-        /*
-         * This is either a duplicate entry or a duplicate entry might follow.
-         * Keep adding unique entries to map and final list
-         */
-        if (!nameToEntryMap.containsKey(next.getPath().getName())) {
-          nameToEntryMap.put(next.getPath().getName(), next);
-          rectifiedFileStatusList.add(next);
-        }
-      } else {
-        // The prefix pattern breaks here.
-        prefix = next.getPath().getName();
-        nameToEntryMap = new TreeMap<>();
-        nameToEntryMap.put(prefix, next);
-        rectifiedFileStatusList.add(next);
-      }
-    }
-
-    return rectifiedFileStatusList;
   }
 
   // generate continuation token for xns account

--- a/hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/utils/ListUtils.java
+++ b/hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/utils/ListUtils.java
@@ -29,6 +29,10 @@ import org.apache.hadoop.fs.FileStatus;
  */
 public class ListUtils {
 
+  private ListUtils() {
+    // Private constructor to prevent instantiation
+  }
+
   /**
    * Utility method to remove duplicates from a list of FileStatus.
    * ListBlob API of blob endpoint can return duplicate entries.

--- a/hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/utils/ListUtils.java
+++ b/hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/utils/ListUtils.java
@@ -19,51 +19,47 @@
 package org.apache.hadoop.fs.azurebfs.utils;
 
 import java.util.ArrayList;
-import java.util.Iterator;
 import java.util.List;
 import java.util.TreeMap;
 
 import org.apache.hadoop.fs.FileStatus;
 
+/**
+ * Utility class for List operations.
+ */
 public class ListUtils {
 
+  /**
+   * Utility method to remove duplicates from a list of FileStatus.
+   * ListBlob API of blob endpoint can return duplicate entries.
+   * @param originalList prone to have duplicates
+   * @return rectified list with no duplicates.
+   */
   public static List<FileStatus> getUniqueListResult(List<FileStatus> originalList) {
     if (originalList == null || originalList.isEmpty()) {
       return originalList;
     }
+
     TreeMap<String, FileStatus> nameToEntryMap = new TreeMap<>();
-    String prefix;
-    Iterator<FileStatus> iterator = originalList.iterator();
+    String prefix = null;
     List<FileStatus> rectifiedFileStatusList = new ArrayList<>();
 
-    FileStatus curr = iterator.next();
-    prefix = curr.getPath().getName();
-    addToUniqueResult(nameToEntryMap, rectifiedFileStatusList, curr);
+    for (FileStatus current : originalList) {
+      String fileName = current.getPath().getName();
 
-    while (iterator.hasNext()) {
-      FileStatus next = iterator.next();
-      if (next.getPath().getName().startsWith(prefix)) {
-        /*
-         * This is either a duplicate entry or a duplicate entry might follow.
-         * Keep adding unique entries to map and final list
-         */
-        if (!nameToEntryMap.containsKey(next.getPath().getName())) {
-          addToUniqueResult(nameToEntryMap, rectifiedFileStatusList, next);
-        }
-      } else {
-        // The prefix pattern breaks here.
-        prefix = next.getPath().getName();
-        nameToEntryMap = new TreeMap<>();
-        addToUniqueResult(nameToEntryMap, rectifiedFileStatusList, next);
+      if (prefix == null || !fileName.startsWith(prefix)) {
+        // Prefix pattern breaks here. Reset Map and prefix.
+        prefix = fileName;
+        nameToEntryMap.clear();
+      }
+
+      // Add the current entry if it is not already added.
+      if (!nameToEntryMap.containsKey(fileName)) {
+        nameToEntryMap.put(fileName, current);
+        rectifiedFileStatusList.add(current);
       }
     }
 
     return rectifiedFileStatusList;
-  }
-
-  private static void addToUniqueResult(TreeMap<String, FileStatus> nameToEntryMap,
-      List<FileStatus> rectifiedFileStatusList, FileStatus fileStatus) {
-    nameToEntryMap.put(fileStatus.getPath().getName(), fileStatus);
-    rectifiedFileStatusList.add(fileStatus);
   }
 }

--- a/hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/utils/ListUtils.java
+++ b/hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/utils/ListUtils.java
@@ -1,0 +1,69 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.fs.azurebfs.utils;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.TreeMap;
+
+import org.apache.hadoop.fs.FileStatus;
+
+public class ListUtils {
+
+  public static List<FileStatus> getUniqueListResult(List<FileStatus> originalList) {
+    if (originalList == null || originalList.isEmpty()) {
+      return originalList;
+    }
+    TreeMap<String, FileStatus> nameToEntryMap = new TreeMap<>();
+    String prefix;
+    Iterator<FileStatus> iterator = originalList.iterator();
+    List<FileStatus> rectifiedFileStatusList = new ArrayList<>();
+
+    FileStatus curr = iterator.next();
+    prefix = curr.getPath().getName();
+    addToUniqueResult(nameToEntryMap, rectifiedFileStatusList, curr);
+
+    while (iterator.hasNext()) {
+      FileStatus next = iterator.next();
+      if (next.getPath().getName().startsWith(prefix)) {
+        /*
+         * This is either a duplicate entry or a duplicate entry might follow.
+         * Keep adding unique entries to map and final list
+         */
+        if (!nameToEntryMap.containsKey(next.getPath().getName())) {
+          addToUniqueResult(nameToEntryMap, rectifiedFileStatusList, next);
+        }
+      } else {
+        // The prefix pattern breaks here.
+        prefix = next.getPath().getName();
+        nameToEntryMap = new TreeMap<>();
+        addToUniqueResult(nameToEntryMap, rectifiedFileStatusList, next);
+      }
+    }
+
+    return rectifiedFileStatusList;
+  }
+
+  private static void addToUniqueResult(TreeMap<String, FileStatus> nameToEntryMap,
+      List<FileStatus> rectifiedFileStatusList, FileStatus fileStatus) {
+    nameToEntryMap.put(fileStatus.getPath().getName(), fileStatus);
+    rectifiedFileStatusList.add(fileStatus);
+  }
+}

--- a/hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemListStatus.java
+++ b/hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemListStatus.java
@@ -586,16 +586,19 @@ public class ITestAzureBlobFileSystemListStatus extends
     // Create Path 9 and 10
     fs.create(new Path("/e/file4"));
 
+    int totalNoOfPaths = 11;
+    int noOfUniquePaths = 7;
+
     FileStatus[] fileStatuses = fs.listStatus(new Path(ROOT_PATH));
 
     // Assert that client.listPath was called 11 times.
     // This will assert server returned 11 entries in total.
-    Mockito.verify(client, Mockito.times(11))
+    Mockito.verify(client, Mockito.times(totalNoOfPaths))
         .listPath(eq(ROOT_PATH), eq(false), eq(1), any(), any(), any());
 
     // Assert that after duplicate removal, only 7 unique entries are returned.
     Assertions.assertThat(fileStatuses.length)
-        .describedAs("List size is not expected").isEqualTo(7);
+        .describedAs("List size is not expected").isEqualTo(noOfUniquePaths);
 
     // Assert that for duplicates, entry corresponding to marker blob is returned.
     assertImplicitDirectoryFileStatus(fileStatuses[0], fs.makeQualified(new Path("/A")));

--- a/hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemListStatus.java
+++ b/hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemListStatus.java
@@ -24,7 +24,9 @@ import java.net.SocketException;
 import java.net.SocketTimeoutException;
 import java.net.URL;
 import java.util.ArrayList;
+import java.util.HashSet;
 import java.util.List;
+import java.util.Set;
 import java.util.concurrent.Callable;
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
@@ -605,12 +607,11 @@ public class ITestAzureBlobFileSystemListStatus extends
     assertExplicitDirectoryFileStatus(fileStatuses[6], fs.makeQualified(new Path("/e")));
 
     // Assert that there are no duplicates in the returned file statuses.
-    for (int i = 0; i < fileStatuses.length; i++) {
-      for (int j = i + 1; j < fileStatuses.length; j++) {
-        Assertions.assertThat(fileStatuses[i].getPath())
-            .describedAs("Duplicate entries found")
-            .isNotEqualTo(fileStatuses[j].getPath());
-      }
+    Set<Path> uniquePaths = new HashSet<>();
+    for (FileStatus fileStatus : fileStatuses) {
+      Assertions.assertThat(uniquePaths.add(fileStatus.getPath()))
+          .describedAs("Duplicate entries found")
+          .isTrue();
     }
   }
 

--- a/hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemListStatus.java
+++ b/hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemListStatus.java
@@ -551,6 +551,7 @@ public class ITestAzureBlobFileSystemListStatus extends
 
     /*
      * Following entries will be created inside the root path.
+     * 0. /A - implicit directory without any marker blob
      * 1. /a - marker file for explicit directory
      * 2. /a/file1 - normal file inside explicit directory
      * 3. /b - normal file inside root
@@ -562,6 +563,8 @@ public class ITestAzureBlobFileSystemListStatus extends
      * 9. /e - marker file for explicit directory
      * 10. /e/file4 - normal file inside explicit directory
      */
+    // Create Path 0
+    createAzCopyFolder(new Path("/A"));
 
     // Create Path 1 and 2.
     fs.create(new Path("/a/file1"));
@@ -583,19 +586,23 @@ public class ITestAzureBlobFileSystemListStatus extends
 
     FileStatus[] fileStatuses = fs.listStatus(new Path(ROOT_PATH));
 
-    // Assert that client.listPath was called 5 times for each entry.
-    Mockito.verify(client, Mockito.times(10))
+    // Assert that client.listPath was called 11 times.
+    // This will assert server returned 11 entries in total.
+    Mockito.verify(client, Mockito.times(11))
         .listPath(eq(ROOT_PATH), eq(false), eq(1), any(), any(), any());
 
-    // Assert that after duplicate removal, 4 entries are returned.
+    // Assert that after duplicate removal, only 7 unique entries are returned.
     Assertions.assertThat(fileStatuses.length)
-        .describedAs("List size is not expected").isEqualTo(6);
-    assertExplicitDirectoryFileStatus(fileStatuses[0], fs.makeQualified(new Path("/a")));
-    assertFilePathFileStatus(fileStatuses[1], fs.makeQualified(new Path("/b")));
-    assertExplicitDirectoryFileStatus(fileStatuses[2], fs.makeQualified(new Path("/c")));
-    assertExplicitDirectoryFileStatus(fileStatuses[3], fs.makeQualified(new Path("/c.bak")));
-    assertImplicitDirectoryFileStatus(fileStatuses[4], fs.makeQualified(new Path("/d")));
-    assertExplicitDirectoryFileStatus(fileStatuses[5], fs.makeQualified(new Path("/e")));
+        .describedAs("List size is not expected").isEqualTo(7);
+
+    // Assert that for duplicates, entry corresponding to marker blob is returned.
+    assertImplicitDirectoryFileStatus(fileStatuses[0], fs.makeQualified(new Path("/A")));
+    assertExplicitDirectoryFileStatus(fileStatuses[1], fs.makeQualified(new Path("/a")));
+    assertFilePathFileStatus(fileStatuses[2], fs.makeQualified(new Path("/b")));
+    assertExplicitDirectoryFileStatus(fileStatuses[3], fs.makeQualified(new Path("/c")));
+    assertExplicitDirectoryFileStatus(fileStatuses[4], fs.makeQualified(new Path("/c.bak")));
+    assertImplicitDirectoryFileStatus(fileStatuses[5], fs.makeQualified(new Path("/d")));
+    assertExplicitDirectoryFileStatus(fileStatuses[6], fs.makeQualified(new Path("/e")));
 
     // Assert that there are no duplicates in the returned file statuses.
     for (int i = 0; i < fileStatuses.length; i++) {

--- a/hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/utils/TestListUtils.java
+++ b/hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/utils/TestListUtils.java
@@ -1,0 +1,86 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.fs.azurebfs.utils;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.assertj.core.api.Assertions;
+import org.junit.Test;
+
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.Path;
+
+public class TestListUtils {
+
+  @Test
+  public void testRemoveDuplicates() {
+    List<FileStatus> originalList = new ArrayList<>();
+    validateList(originalList, 0);
+
+    originalList = new ArrayList<>();
+    originalList.add(getFileStatusObject(new Path("/A")));
+    validateList(originalList, 1);
+
+    originalList = new ArrayList<>();
+    originalList.add(getFileStatusObject(new Path("/A")));
+    originalList.add(getFileStatusObject(new Path("/A")));
+    validateList( originalList, 1);
+
+    originalList = new ArrayList<>();
+    originalList.add(getFileStatusObject(new Path("/a")));
+    originalList.add(getFileStatusObject(new Path("/a.bak1")));
+    originalList.add(getFileStatusObject(new Path("/a.bak1.bak2")));
+    originalList.add(getFileStatusObject(new Path("/a.bak1.bak2")));
+    originalList.add(getFileStatusObject(new Path("/a.bak1")));
+    originalList.add(getFileStatusObject(new Path("/a")));
+    originalList.add(getFileStatusObject(new Path("/abc")));
+    originalList.add(getFileStatusObject(new Path("/abc.bak1")));
+    originalList.add(getFileStatusObject(new Path("/abc")));
+    validateList( originalList, 5);
+
+    originalList = new ArrayList<>();
+    originalList.add(getFileStatusObject(new Path("/a")));
+    originalList.add(getFileStatusObject(new Path("/a")));
+    originalList.add(getFileStatusObject(new Path("/a_bak1")));
+    originalList.add(getFileStatusObject(new Path("/a_bak1")));
+    originalList.add(getFileStatusObject(new Path("/a_bak1_bak2")));
+    originalList.add(getFileStatusObject(new Path("/a_bak1_bak2")));
+    originalList.add(getFileStatusObject(new Path("/abc")));
+    originalList.add(getFileStatusObject(new Path("/abc")));
+    originalList.add(getFileStatusObject(new Path("/abc_bak1")));
+    validateList( originalList, 5);
+  }
+
+  private void validateList(List<FileStatus> originalList, int expectedSize) {
+    List<FileStatus> uniqueList = ListUtils.getUniqueListResult(originalList);
+    Assertions.assertThat(uniqueList)
+        .describedAs("List Size is not as expected after duplicate removal")
+        .hasSize(expectedSize);
+  }
+
+  private FileStatus getFileStatusObject(Path path) {
+    FileStatus status = new FileStatus();
+    status.setPath(path);
+    return status;
+  }
+}

--- a/hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/utils/TestListUtils.java
+++ b/hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/utils/TestListUtils.java
@@ -18,11 +18,8 @@
 
 package org.apache.hadoop.fs.azurebfs.utils;
 
-import java.io.File;
 import java.util.ArrayList;
-import java.util.HashMap;
 import java.util.List;
-import java.util.Map;
 
 import org.assertj.core.api.Assertions;
 import org.junit.Test;
@@ -44,7 +41,7 @@ public class TestListUtils {
     originalList = new ArrayList<>();
     originalList.add(getFileStatusObject(new Path("/A")));
     originalList.add(getFileStatusObject(new Path("/A")));
-    validateList( originalList, 1);
+    validateList(originalList, 1);
 
     originalList = new ArrayList<>();
     originalList.add(getFileStatusObject(new Path("/a")));
@@ -56,7 +53,7 @@ public class TestListUtils {
     originalList.add(getFileStatusObject(new Path("/abc")));
     originalList.add(getFileStatusObject(new Path("/abc.bak1")));
     originalList.add(getFileStatusObject(new Path("/abc")));
-    validateList( originalList, 5);
+    validateList(originalList, 5);
 
     originalList = new ArrayList<>();
     originalList.add(getFileStatusObject(new Path("/a")));
@@ -68,7 +65,18 @@ public class TestListUtils {
     originalList.add(getFileStatusObject(new Path("/abc")));
     originalList.add(getFileStatusObject(new Path("/abc")));
     originalList.add(getFileStatusObject(new Path("/abc_bak1")));
-    validateList( originalList, 5);
+    validateList(originalList, 5);
+
+    originalList = new ArrayList<>();
+    originalList.add(getFileStatusObject(new Path("/a")));
+    originalList.add(getFileStatusObject(new Path("/b")));
+    validateList(originalList, 2);
+
+    originalList = new ArrayList<>();
+    originalList.add(getFileStatusObject(new Path("/a")));
+    originalList.add(getFileStatusObject(new Path("/b")));
+    originalList.add(getFileStatusObject(new Path("/b")));
+    validateList(originalList, 2);
   }
 
   private void validateList(List<FileStatus> originalList, int expectedSize) {

